src/main/java/ch/njol/skript/lang/function/DynamicFunctionReference.java

```diff
@@ -0,0 +1,233 @@
+import ch.njol.skript.lang.Expression;
+import ch.njol.skript.lang.ExpressionList;
+import ch.njol.skript.registrations.Classes;
+import ch.njol.skript.util.Contract;
+import org.bukkit.event.Event;
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+import org.jetbrains.annotations.UnknownNullability;
+import org.skriptlang.skript.lang.script.Script;
+import org.skriptlang.skript.util.Executable;
+import org.skriptlang.skript.util.Validated;
+
+import java.io.File;
+import java.lang.ref.Reference;
+import java.lang.ref.WeakReference;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Objects;
+
+/**
+ * A partial reference to a Skript function.
+ * This reference knows some of its information in advance (such as the function's name)
+ * but will not be resolved until it receives inputs for the first time.
+ * @param <Result> The return type of this function, if known.
+ */
+public class DynamicFunctionReference<Result>
+	implements Contract, Executable<Event, Result[]>, Validated {
+
+	private final @NotNull String name;
+	private final @Nullable Script source;
+	private final Reference<Function<? extends Result>> function;
+	private final @UnknownNullability Signature<? extends Result> signature;
+	private final Validated validator = Validated.validator();
+	private final Map<Input, Expression<?>> checkedInputs = new HashMap<>();
+	private final boolean resolved;
+
+	public DynamicFunctionReference(Function<? extends Result> function) {
+		this.resolved = true;
+		this.function = new WeakReference<>(function);
+		this.name = function.getName();
+		this.signature = function.getSignature();
+		@Nullable File file = ScriptLoader.getScriptFromName(signature.script);
+		if (file == null)
+			this.source = null;
+		else
+			this.source = ScriptLoader.getScript(file);
+	}
+
+	public DynamicFunctionReference(@NotNull String name) {
+		this(name, null);
+	}
+
+	public DynamicFunctionReference(@NotNull String name, @Nullable Script source) {
+		this.name = name;
+		Function<? extends Result> function;
+		if (source != null)
+			//noinspection unchecked
+			function = (Function<? extends Result>) Functions.getFunction(name, source.getConfig().getFileName());
+		else
+			//noinspection unchecked
+			function = (Function<? extends Result>) Functions.getFunction(name, null);
+		this.resolved = function != null;
+		this.function = new WeakReference<>(function);
+		if (resolved) {
+			this.signature = function.getSignature();
+			@Nullable File file = ScriptLoader.getScriptFromName(signature.script);
+			if (file == null)
+				this.source = null;
+			else
+				this.source = ScriptLoader.getScript(file);
+		} else {
+			this.signature = null;
+			this.source = null;
+		}
+	}
+
+	public @Nullable Script source() {
+		return source;
+	}
+
+	public String name() {
+		return name;
+	}
+
+	@Override
+	public boolean isSingle(Expression<?>... arguments) {
+		if (!resolved)
```

Comment:
`resolved` or `valid()`? Would it be possible to combine the two for clarity? Is a resolved function ever invalid? That is, can we accurately return this for an invalid(ated) reference?

Reply:
Valid functions are always resolved, resolved functions are not always valid.
A Dynamic function becomes resolved once the actual, underlying `Function` is linked to it, but it's valid only if that function is actually valid itself (i.e. it's loaded, the script its from is still valid, hasn't been reloaded, etc.) so the validity could change.
Validity is slightly more of a complex check so I avoided it here because I'm just getting some details from the function, though ideally I should keep it consistent with below and make them both use resolved.